proposals/0448-type-variable-scoping.rst

```diff
@@ -0,0 +1,1167 @@
+   to the following constructs:
+
+   * Function bindings
+   * Pattern synonym bindings (including in any ``where`` clause)
+   * Expression type signatures
+
+   (Alternative name: ``-XExtendedForAllScope``. See `Vlad's comment <https://github.com/ghc-proposals/ghc-proposals/pull/448#discussion_r738276607>`_.).
+
+   Separating out ``-XScopedForAlls`` gets us closer to the `Contiguous Scoping Principle`_.
+
+#. The extension ``-XScopedTypeVariables`` would imply all of the above
+   extensions; this way, ``-XScopedTypeVariables`` does not change from its
+   current meaning.
+
+#. Introduce ``-XImplicitForAll``, on by default. With ``-XImplicitForAll``,
+   an out-of-scope type variable mentioned in various constructs (listed below)
+   is implicitly brought into scope over the construct. With ``-XNoImplicitForAll``,
+   this implicit scoping does not happen, and the use of the variable is an error.
+
+   Constructs affected:
+
+   1. Type signatures for variable declarations, methods, and foreign imports & exports.
+      Example: ``let f :: a -> a; f = ... in ...`` becomes
+      ``let f :: forall a. a -> a; f = ... in ...``.
+
+   #. Kind signatures. Example: ``type T :: k -> Type`` becomes ``type T :: forall k. k -> Type``.
+
+   #. GADT constructor declarations. Example: ``MkG :: a -> Maybe b -> G (Either Int b)``
+      becomes ``MkG :: forall a b. a -> Maybe b -> G (Either Int b)``.
+
+   #. Pattern synonym signatures. Example: ``pattern P :: a -> Maybe a`` becomes
+      ``pattern P :: forall a. a -> Maybe a``. Implicit quantification in pattern synonyms
+      always produces *universal* variables, never existential ones.
+
+   #. Type annotations in expressions and ``SPECIALISE`` pragmas. Example:
+      ``Right True :: Either a Bool`` becomes ``Right True :: forall a. Either a Bool``.
+
+   #. Types in a ``deriving`` clause. Example: ``data T deriving (C a)`` becomes
+      ``data T deriving (forall a. C a)``.
+
+   #. Instance heads, including standalone-deriving instances.
+      Example: ``instance Show a => Show (Maybe a)`` becomes
+      ``instance forall a. Show a => Show (Maybe a)``.
+
+   #. Type and data family instances, as well as closed type family equations.
+      Example: ``type instance F (Maybe a) = Int``
+      becomes ``type instance forall a. F (Maybe a) = Int``.
+
+   #. ``RULES`` pragmas.
+      Example: ``{-# RULES "name" forall (x :: Maybe a). foo x = 5 #-}``
+      becomes ``{-# RULES "name" forall a. forall (x :: Maybe a). foo x = 5 #-}``.
+      (The double-\ ``forall`` syntax separates type variables like ``a`` from
+      term variables like ``x``.)
+
+   This extension is part of accepted, unimplemented proposal `#285`_; the only change is including ``RULES`` pragmas, which @Ericson2314 simplify forgot to include in `#285`_ (his own admission).
+
+   Being able to turn off this extension is necessary to uphold the `Explicit Binding Principle`_.
```

Comment:
I'll insist that `-XImplicitForAll` is actually unrelated to the rest of the proposal and should just be removed. And we can just merge #480 to fix the infelicity with the `RULES` pragma.

```suggestion
```

Reply:
`-XImplicitForAll` *is* related: it is part of what gives us the Explicit Binding Principle. I mean, removing it from this proposal doesn't really change anything (because it's already accepted), but this is part of what makes this proposal present the "big picture".